feat: Add UI_EVENT PLAY, PLAY_ARTIST,  PLAY_GENRE, PLAY_ALBUM_ARTIST from overflow

BumbleFlash opened the pull request:
This PR is to handle the case when the user clicks `PLAY` from the overflow button.

Changes include

- [x] Capture UI_EVENT `PLAY` from Overflow button in `AlbumArtistFragment`.
- [x] Capture UI_EVENT `PLAY` from Overflow button in `ArtistFragment`.
- [x] Capture UI_EVENT `PLAY` from Overflow button in `SuggestedFragment`
- [x] Capture UI_EVENT `PLAY` from Overflow button in `ArtistFragment`
- [x] Capture UI_EVENT `PLAY` from Overflow button in `GenreFragment`
- [x] Move `getSong` to `MusicServiceConnectionUtils.getSong()` as discussed in https://github.com/CUTR-at-USF/MUSER/pull/90#discussion_r510475904
- [x] Create data class `AlbumArtistdata` and `ArtistData`
- [x] Add `album` and artist` fields to the root of the model.
- [x] Populate the fields respectively while the corresponding overflow is clicked.
- [x] Add PLAY_ARTIST and PLAY_ALBUM
- [x] Use AlbumData to record the List of Albums in the `PLAY_ALBUM_ARTIST` event.  
- [x] Add PLAY_GENRE event.  
- [x] Move `getSong` and `isPlaying` to `MusicServiceConnectionUtils`
- [x] Remove `?` from `newPlayerEvent` function and add null checks. 

"Closes #89", part of #17

BumbleFlash committed:
chore: added UI_EVENT from AlbumArtistFragment and AlbumFragment

BumbleFlash committed:
chore: added UI_EVENT from GenreListFragment.

BumbleFlash commented:
@barbeau I have added a delay of 200ms after the overflow button is clicked to capture UI_EVENT `PLAY` in order to give time for the `MusicService` to update with the current song. I wasn't sure how else to wait for the async function to complete so I came up with this.


      Timer().schedule(200) {
                  newUiEvent(UiEventType.PLAY)
                }

Is there a neater way to do this? or should we keep this as is? Thank you!

barbeau commented:
@BumbleFlash Hmmm...that's definitely not ideal since we don't have any guarantee how long it will take the `MusicService` to update state. So we could log the UI event for the wrong song if the UI and service are out of sync.

If the user is tapping on the song description in the UI, then the UI must have some knowledge of what that song is to show the description to the user. It may not be as complete as the service, but I would immediately log the UI event with the song info knowledge that the UI has. Ideally if the song info in the UI isn't complete we'd somehow query the service for the song info to fill in the blanks - this could be async. Not sure how simple that is given the current app structure. 

But if it's between immediately logging partial song info from the UI, versus having to schedule a timer to log the event in the near future (when we don't know for sure that `MusicService` will have the proper song info), I'd err on the side of immediately logging the partial song info.

barbeau commented:
Remember, even if we log a UI PLAY event with partial song info, we should log a player PLAY event at approx. the same time with the complete info - so it shouldn't be that hard to fill in server-side either.

BumbleFlash commented:
@barbeau Thanks for your input. 
To keep it brief, Shuttle uses custom `Presenter` classes to handle callbacks that are used in `MenuUtils`. Each Fragment has a custom `Presenter` class and a `MenuUitls` class. When the overflow button is clicked, say from the `AlbumFragment`, the argument of the callback is the model `Album` class. So, I'm guessing we would only be able to collect the Album name for the specific example. Also, there's no `Play` overflow button on a Song, it's only on the `Album` or `AlbumArtist`.

barbeau commented:
@BumbleFlash Gotcha, thanks for the info. So I think in those cases we just want to log that a user specified via the UI that they wanted to play a particular album or a particular artist - so just logging that info as a UI event is fine. I think it would actually be somewhat deceptive to log a particular song as a UI event in those cases as the user didn't actually choose a particular song to play.

A question for us is if we want to expand the `UiEvent` class with additional fields to handle this case instead of stuffing the data into the `song` field. `song` is already optional, so we could add optional `album` and `albumArtist` fields as well. It seems like this model would give us a clearer indication of what type of data was available when the event was logged. Thoughts?

BumbleFlash commented:
Yes. Sounds like a great idea. It would help us maintain consistency in the attributes, especially when we have the data in a relational manner. I'm guessing while logging the "PLAY" from the overflow button, we would only populating the `album` or `albumartist` fields, right? That means we would have empty fields on SongData. So I'm thinking that we should have default values populated for all the variables in the model now, in case they were null?

barbeau commented:
>I'm guessing while logging the "PLAY" from the overflow button, we would only populating the album or albumartist fields, right? 

Correct

>That means we would have empty fields on SongData. So I'm thinking that we should have default values populated for all the variables in the model now, in case they were null?

`song` is already nullable, right? So I think we can leave as-is. I anticipated that we would have UI events that wouldn't include a reference to a song (e.g., clicking on buttons that don't produce player events).

BumbleFlash committed:
chore: added fields album and albumartist and converted EventUtils.kt to kotlin file

BumbleFlash committed:
chore: defined two functions to capture album and album artist data

BumbleFlash committed:
chore: added PLAY_ALBUM and PLAY_ALBUM_ARTIST to the UiEventType

BumbleFlash committed:
chore: captured UI_EVENT PLAY_ALBUM and PLAY_ALBUM_ARTIST and fixed compile errors

BumbleFlash commented:
@barbeau one new addition to the function which needed to be done. I have created two data classes, `AlbumData` and `AlbumArtistData` to avoid improper serialization errors. The `AlbumData` seems to be doing well, however, in `AlbumArtistData`, I had to remove the `List<Album> albums` property because `Album` class from Shuttle was throwing a "No properties to serialize found" error. I'm guessing the reason behind this is the private properties in the `Album` class namely `artworkKey` and `sortKey`. https://github.com/CUTR-at-USF/MUSER/blob/usf/app/src/main/java/com/simplecity/amp_library/model/Album.java#L42

BumbleFlash commented:
@barbeau Also, there's a bit of a lag while clicking the `PLAY` from the overflow button, currently figuring out the reason behind it.

BumbleFlash committed:
chore: removed extra space

BumbleFlash commented:
Update: I can quite confirm the lag is not due to this PR after testing the MUSER `1.1.2` for the same. I also tested the latest version of Shuttle and the lag seems to be present on there as well.

BumbleFlash committed:
chore: renamed SongUtils.java to MusicServiceUtils.java and moved getSong to MusicServiceUtils

BumbleFlash commented:
@barbeau I think the Github gradle is failing to build the project because of the change in the implementation of the `EventUtils.kt`  in this PR and how it would potentially affect the current `usf` branch. The CLI is failing in the `PlayerPresenter.kt` possibly where the `NEXT` event is being recorded in the `usf` branch.

BumbleFlash committed:
Merge branch 'usf' into ui-event-overflow

BumbleFlash commented:
@barbeau Is this okay to merge the current `usf` branch to solve the compile error and then merge it back to `usf`? I hope  I didn't go ahead of the game and did something that I'm not supposed to do

BumbleFlash committed:
chore: fixed compile error due to the change in implementation of EventUtils.kt

BumbleFlash committed:
chore: Used Albumdata instead of Album in the AlbumArtistdata class to record album information

barbeau commented:
@BumbleFlash I'm not sure I understand your question. All changes in this PR should be self-contained, so unless the `usf` branch is failing to build you should be able to change the code in this branch to compile correctly. It looks like it's passing the tests on CI?
https://github.com/CUTR-at-USF/MUSER/runs/1315793907

BumbleFlash commented:
@barbeau Yes - you're right. It compiles now because I merged the `usf` branch to the PR locally and fixed a compile error that raised after the conversion of `Eventutils` to **kotlin** file and changing one of the `newUiEvent` function arguments from `song?` to `song`. You could see this run over here https://github.com/CUTR-at-USF/MUSER/runs/1315434689 that wasn't compiling before I merged `usf` locally. What I'm curious about is, before I fixed the error, if you haven't noticed, the reason why it failed to compile was a type mismatch error in the `newUiEvent` function in `PlayerPresenter.kt` in the `usf` branch. I can quite confirm that I never touched this file in this PR. As I fixed the error, the GitHub CI compiles it successfully.

BumbleFlash commented:
@barbeau Also, FYI, I think I have completed all the checklists and it's ready for your review. :)

barbeau commented on app/src/main/java/com/simplecity/amp_library/ui/screens/genre/menu/GenreMenuPresenter.kt:
```diff
@@ -45,6 +52,9 @@ class GenreMenuPresenter @Inject constructor(
         mediaManager.playAll(genre.getSongs(context)) {
             view?.onPlaybackFailed()
         }
+        Timer().schedule(200) {
```
We shouldn't need this `Timer` anymore, right?

BumbleFlash replied:
No, we shouldn't be needing it after I record the `PLAY_GENRE` UI_EVENT.

barbeau commented on app/src/main/java/com/simplecity/amp_library/ui/screens/nowplaying/PlayerPresenter.kt:
```diff
@@ -302,8 +303,8 @@ class PlayerPresenter @Inject constructor(
-        if (PlayerFragment.getSong() != null) {
-            val uiEvent = EventUtils.newUiEvent(PlayerFragment.getSong(), uiEventType, context)
+        if (MusicServiceUtils.getSong() != null) {
+            val uiEvent = EventUtils.newUiEvent(MusicServiceUtils.getSong()!!, uiEventType, context)
```
You can avoid the `!!` here by doing something like this:

```kotlin
Song song = MusicServiceUtils.getSong()
 if (song  != null) {
            val uiEvent = EventUtils.newUiEvent(song, uiEventType, context)
...
```

...because you'd guaranteeing that the `song` variable won't be `null` at that point. That's better than potentially having the state change between the two invocations of `MusicServiceUtils.getSong()` which could potentially end up throwing an NPE at the 2nd invocation.

BumbleFlash replied:
Sounds good. Thank you for letting me know. I'll make it similar for other `!!` in the code.

barbeau commented on app/src/main/java/com/simplecity/amp_library/ui/screens/songs/list/SongListFragment.kt:
```diff
@@ -288,7 +288,7 @@ class SongListFragment :
     override fun onSongClick(position: Int, songView: SongView) {
         if (!contextualToolbarHelper!!.handleClick(songView, songView.song)) {
-            val uiEvent = EventUtils.newUiEvent(songView.song, UiEventType.PLAY, context)
+            val uiEvent = EventUtils.newUiEvent(songView.song, UiEventType.PLAY, context!!)
```
For this and another invocations of the same method - if `context` can't be null, it's better to do a null check before executing this to avoid potentiall issues with null contexts. So:

```kotlin
if (context != null) {
    val uiEvent = EventUtils.newUiEvent(songView.song, UiEventType.PLAY, context)
    ...
````

`!!` will throw an NPE if `context` is null, so if we want to avoid NPEs we really shouldn't be using `!!`. See https://kotlinlang.org/docs/reference/null-safety.html.

barbeau commented on app/src/main/java/edu/usf/sas/pal/muser/model/AlbumData.kt:
```diff
@@ -0,0 +1,20 @@
+ val dateAdded: Long,
+ val paths: List<String>
+ ){
+    constructor(album: Album): this(album.id, album.name, album.artists, album.albumArtistName,
```
I believe you should be able to get rid of these explicit `constructor`s in the data classes if you're calling exclusively from Kotlin code and using named parameters.

BumbleFlash replied:
Currently, I'm using this constructor to populate the `AlbumData` class and it doesn't use any named parameters. Also, the named parameters are only being used to populate the `UiEvent` class.

barbeau replied:
Ah, ok, thanks. Sorry, I read this too quickly the first time 👍

barbeau commented on app/src/main/java/edu/usf/sas/pal/muser/util/EventUtils.kt:
```diff
@@ -0,0 +1,99 @@
+     * @return PlayerEvent object
+     */
+    @JvmStatic
+    fun newPlayerEvent(song: Song?, capturedEvent: PlayerEventType?, context: Context?): PlayerEvent {
```
Let's try to remove the `?` from `PlayerEventType` and `Context` - I believe `Song` should be the only nullable parameter here for player events that don't have songs. (is that right, or are we assuming that all player events have songs now?)

barbeau replied:
That should also get rid of the `!!` below. We want to avoid those.

BumbleFlash replied:
I think it's safe to assume that `song` is always populated since we're logging at the lowest level of the service as possible. But, if we wanna add an extra layer of protection just in case, then I can go ahead and remove the `?` and add null checks.

barbeau replied:
Looking at the code, if I change to this (removing all `?` and `!!`) it builds and runs:

```kotlin
    @JvmStatic
    fun newPlayerEvent(song: Song, capturedEvent: PlayerEventType, context: Context): PlayerEvent {
        val currentTimeMS = System.currentTimeMillis()
        val nanoTime = System.nanoTime()
        val songData = SongData(song, context)
        val seekPositionMs = position
        return PlayerEvent(capturedEvent, currentTimeMS, nanoTime, seekPositionMs,
                songData)
    }
```

Let's do that. `song` can potentially be `null` (looking at the Java classes it's called from), but I think we actually want to crash in this case instead of not logging the event so that we know we are missing events via crash logs in Google Play. As long as there isn't a valid use case where we expect `song` to be `null` - it sounds like there isn't.

barbeau commented on app/src/main/java/edu/usf/sas/pal/muser/util/MusicServiceUtils.java:
```diff
@@ -0,0 +1,24 @@
+import com.simplecity.amp_library.model.Song;
+import com.simplecity.amp_library.utils.MusicServiceConnectionUtils;
+
+public class MusicServiceUtils {
```
Is there any reason not to move the below two methods to `MusicServiceConnectionUtils` instead of this new class?

If we need a new class, I'd prefer it to be in Kotlin rather than Java.

BumbleFlash replied:
I was thinking if we have our utils class of our own in our package `edu.usf.sas.pal.muser`, it'd be easier for us to differentiate between our code and the legacy "Shuttle" code in the long run. And yes, I can covert this to kotlin in the future commit. But, feel free to let me know if you would like to me move these to the original `MusicServiceConnectionUtils`.

barbeau replied:
Let's go ahead and move to `MusicServiceConnectionUtils`. It seems like these methods should belong there anyway.

barbeau reviewed (changes requested):
@BumbleFlash Thanks! This looks pretty good - a few comments in-line.

BumbleFlash commented:
@BumbleFlash After looking at the `GenreMenuPresenter.kt`, I think we are going to need a new UI_EVENT `PLAY_GENRE` when the user plays a song from the "Genre" tab? I'll have them ready in the future commits if that's okay with you. Let me know what you think.

BumbleFlash committed:
chore: added condition check to remove "!!" operator

barbeau commented:
>After looking at the GenreMenuPresenter.kt, I think we are going to need a new UI_EVENT PLAY_GENRE when the user plays a song from the "Genre" tab?

👍

BumbleFlash committed:
chore: created model Genre and added PLAY_GENRE event

BumbleFlash committed:
chore: moved getSong and isPlaying to MusicServiceConnectionUtils

BumbleFlash committed:
chore: Removed "?" from newPlayerEvent and added null checks

barbeau commented on app/src/main/java/com/simplecity/amp_library/playback/PlaybackManager.java:
```diff
@@ -548,8 +548,10 @@ public void play() {
      void newPlayerEvent(Song song, PlayerEventType playerEventType){
-        PlayerEvent playerEvent = EventUtils.newPlayerEvent(song, playerEventType, ShuttleApplication.get());
-        FirebaseIOUtils.savePlayerEvent(playerEvent);
+        if (song != null) {
```
I'd remove the null check here because we want to know if we're missing logging events due to null songs. If it crashes, we'll know via the Google Play console and then we can fix it.

barbeau commented on app/src/main/java/com/simplecity/amp_library/playback/QueueManager.java:
```diff
@@ -147,8 +147,10 @@ public void load(@NonNull List<Song> songs, final int position, @NonNull UnsafeA
     void newPlayerEvent(Song song, PlayerEventType playerEventType){
-        PlayerEvent playerEvent = EventUtils.newPlayerEvent(song, playerEventType, ShuttleApplication.get());
-        FirebaseIOUtils.savePlayerEvent(playerEvent);
+        if (song != null) {
```
Same - please remove null check

barbeau commented on app/src/main/java/com/simplecity/amp_library/ui/screens/album/detail/AlbumDetailFragment.kt:
```diff
@@ -403,9 +403,12 @@ class AlbumDetailFragment :
-            FirebaseIOUtils.saveUiEvent(uiEvent)
             if (!contextualToolbarHelper!!.handleClick(songView, Single.just(listOf(songView.song)))) {
+                val context = context
+                if (context != null) {
```
Same - please remove null check

BumbleFlash replied:
When I remove the conditions, AS throws a `Typemismatch` error since `newUiEvent` takes in `context` instead of `context?`. However, it is not the same case for the `song` object.

barbeau replied:
You can pass in `context!!` instead - that will throw an NPE if `context` is null.

However - let's use `ShuttleApplication.get()` instead of the `context` variable of the fragment. That will use the application context, which should always exist as long as the code is running, which is fine as the context is only used to get the string value from resources.

barbeau commented on app/src/main/java/com/simplecity/amp_library/ui/screens/artist/detail/ArtistDetailFragment.kt:
```diff
@@ -154,9 +154,12 @@ class ArtistDetailFragment :
-            FirebaseIOUtils.saveUiEvent(uiEvent)
             if (!contextualToolbarHelper!!.handleClick(songView, Single.just(listOf(songView.song)))) {
+                val context = context
+                if (context != null) {
```
Same - please remove null check

BumbleFlash replied:
Same here. AS throws an error.

barbeau commented on app/src/main/java/com/simplecity/amp_library/ui/screens/genre/detail/GenreDetailFragment.kt:
```diff
@@ -450,9 +450,12 @@ class GenreDetailFragment :
-                FirebaseIOUtils.saveUiEvent(uiEvent)
-                presenter.songClicked(songView.song)
+                val context = context
+                if (context != null) {
```
Same - remove null check

BumbleFlash replied:
Same here. AS throws an error.

barbeau commented on app/src/main/java/com/simplecity/amp_library/ui/screens/nowplaying/PlayerPresenter.kt:
```diff
@@ -302,8 +303,9 @@ class PlayerPresenter @Inject constructor(
-            val uiEvent = EventUtils.newUiEvent(PlayerFragment.getSong(), uiEventType, context)
+        val song = MusicServiceConnectionUtils.getSong()
+        if (song != null) {
+            val uiEvent = EventUtils.newUiEvent(song, uiEventType, context)
```
Same for all cases where logging events - remove null check

BumbleFlash replied:
Same here. AS throws an error, because the argument type is `song` instead of `song?`

barbeau replied:
I think this is a good use case for `song!!`, assuming that we believe that `song` should never be `null` here.

barbeau commented on app/src/main/java/edu/usf/sas/pal/muser/util/EventUtils.kt:
```diff
@@ -0,0 +1,115 @@
+     * function to get the seek position of the track.
+     * @return current seek position. 0 if there's an error
+     */
+    val position: Long
```
Could you move this method to MusicServiceConnectionUtils as well? It really should also be in the form of a `fun xxx(` and not a variable with a `get()` function within.

barbeau reviewed (changes requested):
@BumbleFlash Thanks, this is almost done! See a few of my comments within. Basically, we want to know if we're missing events due to null values, so we shouldn't null check them prior to passing into the Kotlin utility methods.

BumbleFlash committed:
chore: moved getPosition to MusicServiceConnectionUtils

BumbleFlash committed:
chore: removed null checks for the song object

BumbleFlash committed:
chore: passed ShuttleApplication.get() instead of context

BumbleFlash committed:
chore: removed song null checks and added "!!"

barbeau reviewed (approved):
LGTM! Thanks @BumbleFlash!
